### DISCUSSION TITLE
Is `machine.transition()` a pure function?

### DISCUSSION THREAD

**Discussion by Leftium** [statelyai/xstate#4001] [Q&A] [answered]
According to the [docs](https://stately.ai/docs/xstate/running-machines/intro#running-machines-as-pure-functions:~:text=You%20can%20use%20machine.transition%20to%20run%20your%20machine%20as%20a%20pure%20function%20without%20executing%20any%20of%20its%20actions.):

> You can use `machine.transition` to run your machine as a pure function without executing any of its actions.

However, according to [this test case](https://github.com/statelyai/xstate/blob/d98d948a534b028f4af8b271eb431a8625ab5cb3/packages/xstate-immer/test/immer.test.ts#L4-L31), `context` is expected to be updated by the Immer `assign` action. So the action must have been executed:

    const zeroState = countMachine.initialState;
    const oneState = countMachine.transition(zeroState, 'INC');
    const twoState = countMachine.transition(zeroState, 'INC');

    expect(zeroState.context).toEqual({ count: 0 });
    expect(oneState.context).toEqual({ count: 1 });
    expect(twoState.context).toEqual({ count: 1 });

I originally interpreted the docs as "the actions are not run at all." However do the docs claim `machine.transition` is pure because the original machine is not touched and the `assign` action is run on a new machine?

That's actually the behavior I want; just trying to confirm if my understanding is correct~ This will be true for vanilla `assign`, as well, right?

**update:**

If the immer test case above were used with this [custom interpreter](https://xstate.js.org/docs/guides/interpretation.html#custom-interpreters), wouldn't that mean the actions are run twice?
- First by `machine.transition()`
- Then by the custom interpreter?

**Comment by davidkpiano** [comment, marked as answer]
Yes, so `machine.transition()` will determine the next state, which includes `state.value` and `state.context`, but not execute non-built-in actions, so it is a pure function.

This is why `assign(...)` and `raise(...)` will be "executed" when determining the next state; they're not actually side-effects since they only influence the resulting next state and not anything external.

**Comment by Leftium** [comment]
Thanks for the explanation! Perhaps the documentation could make this more clear?

Also, doesn't this mean the sample [custom interpreter](https://xstate.js.org/docs/guides/interpretation.html#custom-interpreters) will execute the `assign(...)` and `raise(...)` actions twice? (First by `machine.transition()`, then by the custom interpreter.) If so, how could you execute only non-built-in actions (or are built-in actions not included because they were already executed?)
